sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InlineCTE.scala

```diff
@@ -71,53 +71,61 @@ case class InlineCTE(
-   *                       bookkeeping work.
+   * @param outerCTEId While collecting the map we use this optional CTE id to identify the
+   *                   current outer CTE.
    */
   private def buildCTEMap(
       plan: LogicalPlan,
       cteMap: mutable.Map[Long, CTEReferenceInfo],
-      collectCTERefs: CTERelationRef => Unit = _ => ()): Unit = {
+      outerCTEId: Option[Long] = None): Unit = {
     plan match {
       case WithCTE(child, cteDefs) =>
+        var visit = true
         cteDefs.foreach { cteDef =>
-          cteMap(cteDef.id) = CTEReferenceInfo(
-            cteDef = cteDef,
-            refCount = 0,
-            outgoingRefs = mutable.Map.empty.withDefaultValue(0),
-            shouldInline = true
-          )
+          cteMap.get(cteDef.id) match {
+            // If we have seen this CTE definition then it must be self-contained so we can clear
+            // the references from its container to the definition
+            case Some(refInfo) =>
+              refInfo.container.foreach(c => cteMap(c).outgoingRefs -= cteDef.id)
+
+              // We have processed this definition (and all others in this `WithCTE` node) so we
+              // don't need to do it again.
+              visit = false
+
+            case _ =>
+              cteMap(cteDef.id) = CTEReferenceInfo(
+                cteDef = cteDef,
+                refCount = 0,
+                outgoingRefs = mutable.Map.empty.withDefaultValue(0),
+                shouldInline = true,
+                container = outerCTEId
+              )
+          }
         }
-        cteDefs.foreach { cteDef =>
-          buildCTEMap(cteDef, cteMap, ref => {
-            // A CTE relation can references CTE relations defined before it in the same `WithCTE`.
-            // Here we update the out-going-ref-count for it, in case this CTE relation is not
-            // referenced at all and can be optimized out, and we need to decrease the ref counts
-            // for CTE relations that are referenced by it.
-            if (cteDefs.exists(_.id == ref.cteId)) {
-              cteMap(cteDef.id).increaseOutgoingRefCount(ref.cteId, 1)
-            }
-            // Similarly, a CTE relation can reference CTE relations defined in the outer `WithCTE`.
-            // Here we call the `collectCTERefs` function so that the outer CTE can also update the
-            // out-going-ref-count if needed.
-            collectCTERefs(ref)
-          })
+
+        if (visit) {
+          cteDefs.foreach { cteDef =>
+            buildCTEMap(cteDef, cteMap, Some(cteDef.id))
+          }
+          buildCTEMap(child, cteMap, outerCTEId)
         }
-        buildCTEMap(child, cteMap, collectCTERefs)
 
       case ref: CTERelationRef =>
         cteMap(ref.cteId) = cteMap(ref.cteId).withRefCountIncreased(1)
-        collectCTERefs(ref)
+        outerCTEId.foreach { cteId =>
+          cteMap(cteId).increaseOutgoingRefCount(ref.cteId, 1)
```

Comment:
Can we add some comments here? We should explain that the `outerCTEId` may not really reference this CTE relation, but contains. However, we treat them the same as there is no harm in updating the reference count even if it's a "contains" relationship.

Reply:
Yeah, good idea. Done in https://github.com/apache/spark/pull/49379/commits/c62c6c4bf2ec4591fee1c5129315a2d9878159a9.